cmd/gapit/framegraph.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/google/gapid/core/app"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+)
+
+// Default output file name.
+const defaultOutFilename = "framegraph.dot"
+
+type framegraphVerb struct{ FramegraphFlags }
+
+func init() {
+	verb := &framegraphVerb{}
+	app.AddVerb(&app.Verb{
+		Name:      "framegraph",
+		ShortHelp: "Create frame graph (in DOT format) from capture",
+		Action:    verb,
+	})
+}
+
+// Run is the main logic for the 'gapit framegraph' command.
+func (verb *framegraphVerb) Run(ctx context.Context, flags flag.FlagSet) error {
+	if flags.NArg() != 1 {
+		app.Usage(ctx, "Exactly one gfx trace file expected, got %d", flags.NArg())
+		return nil
+	}
+
+	captureFilename := flags.Arg(0)
+
+	client, capture, err := getGapisAndLoadCapture(ctx, verb.Gapis, GapirFlags{}, captureFilename, CaptureFileFlags{})
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	boxedFramegraph, err := client.Get(ctx, capture.Framegraph().Path(), nil)
+	if err != nil {
+		return err
+	}
+	framegraph := boxedFramegraph.(*api.Framegraph)
+
+	dot := framegraph2dot(framegraph, captureFilename)
+
+	// Write the DOT representation of the framegraph into a file
+	filePath := verb.Out
+	if filePath == "" {
+		filePath = defaultOutFilename
+	}
+	file, err := os.Create(filePath)
+	if err != nil {
+		return log.Errf(ctx, err, "Creating file (%v)", filePath)
+	}
+	defer file.Close()
+	bytesWritten, err := fmt.Fprint(file, dot)
+	if err != nil {
+		return log.Errf(ctx, err, "Error after writing %d bytes to file", bytesWritten)
+	}
+
+	return nil
+}
+
+// framegraph2dot formats a framegraph in the Graphviz DOT format.
+// https://graphviz.org/doc/info/lang.html
+func framegraph2dot(framegraph *api.Framegraph, captureFilename string) string {
```

Comment:
A general feedback about this: 

Golang strings are immutable and every time there is a an add operation, it may trigger a new allocation. So, this does not seem to be a hot path call, it's probably a call that runs once per query so not a big issue but if you plan to do more work on this in the future, it may be good to look at if this can be converted to something easier to read, more efficient to run with either string formatting or other alternatives.

Reply:
Thanks for the feedback!

I am aware of the perf impact of the string manipulations here, I used it for prototyping and knowing this won't be noticeable given that we deal with not-very-big files here. (To give an idea, generating the dependency graph is in the order of several seconds, so in comparison producing the framegraph (which is much smaller than the dependency graph) dot representation is negligible.)

I'll keep your remarks in mind and improve once we have a working framegraph!